treesitter: Clip injected ranges to the parent LanguageTree's ranges

cormacrelf opened the pull request:
Fixes #21309, fixes #18176

> The idea is to take the injected `@rust` range `{ 3, 4, 5, 0 }` and 'clip' or 'hole punch' the bit that isn't allowed, which splits it into two ranges `{3, 4, 4, 0}`, `{4, 4, 5, 0}`.

This is the solution in a nutshell:

1. ~~Write a centred interval tree in Lua based on the [Wikipedia outline](https://en.wikipedia.org/wiki/Interval_tree).~~ Instead, write a binary search for ranges in a region that overlap with a given range. Regions are sorted.
2. Make it work with the 6-length { start_row, start_col, start_byte, end_row, end_col, end_byte } range format.
3. A function to take any injection range, and intersect it with a list of ranges that overlap with it, producing a list of clipped ranges.
4.  Then, add an argument to `LanguageTree:set_included_regions` to supply the _parent_ LanguageTree's `_regions`.
5. In `set_included_regions` we
    * ~~Flatten the parent's regions into one long region~~ match the child region with the parent region
    * ~~Use that region to construct a new interval tree~~
    * For each of the child's new regions, for each range in it, clip the range against the parent ~~interval tree~~ region.

An interval tree or similar structure seems necessary here because when the injection query matches are returned from tree-sitter, they have lost the information about which of the included region(s) they cover. I think you could potentially add something to the TS C API to get information about this (i.e. access to indexes of overlapping regions that TS could store with nodes as it parses) to avoid having to reconstruct the information out of band at a cost. That being said, this seems to work pretty well and does not seem to be a big slowdown. A file with 5000 Rust doc comments is equally (dreadfully) slow with or without this PR.

cormacrelf committed:
fix(test): exclude __uint128_t from Darwin cdef headers

clason commented:
Thank you for tackling this! I believe that correctness > performance here (keeping in mind that performance can be tackled in a follow-up when we can test against regressions).

Speaking of: a test would be nice :)

I also see that you formalized the `Range` parameter. This would be good to have consistently across the treesitter API (where we have a number of range utility functions, some of which are conflated with nodes and should be untangled).

cormacrelf commented:
I can probably use those three toy languages to make a test for this behaviour. It seems I could modify the CMake TS fetcher to point to a "vendored" copy of them, i.e. the grammar.js, src/parser.c etc files somewhere in the tree. Where would they go?

And yeah it's a distinction in name only, Range6 is just number[]. I think you could do better with `@alias Range6 { [1]: number, [2]: number, ... }` but haven't tried yet. Is there a policy on using LuaJIT FFI types to avoid using tables altogether? I think you can get some crazy performance out of that for fixed-size table-twiddling code, but this is largely dynamic sized tables so probably not.

clason commented:
> I can probably use those three toy languages to make a test for this behaviour. It seems I could modify the CMake TS fetcher to point to a "vendored" copy of them, i.e. the grammar.js, src/parser.c etc files somewhere in the tree. Where would they go?

Hmm. That sounds a bit too heavy for a functional test. Can't it be mocked up? We just want to guard against regression.

> And yeah it's a distinction in name only, Range6 is just number[]. I think you could do better with @alias Range6 { [1]: number, [2]: number, ... } but haven't tried yet. 

Sure; it's more about consistent use than any actual class/object structure.

> Is there a policy on using LuaJIT FFI types to avoid using tables altogether?

Yes: Don't. We only require PUC Lua, not LuaJIT, so you can't assume it's available. (Unfortunately.)

>  I think you can get some crazy performance out of that for fixed-size table-twiddling code, but this is largely dynamic sized tables so probably not.

Could be worth benchmarking though, if it's easy to make optional.

cormacrelf commented:
I can't see a mock surviving much in terms of alternative implementations, unless I mock quite deep. Think moving the clipping to iter_matches, if that e.g. fixes some highlight problems for everyone at once. A perennial issue with mocks I suppose. I'll give it a go.

clason commented:
Understood. Doesn't have to be a full test, and if it doesn't make sense, then that's not a blocker.

cormacrelf commented:
I've realised a couple of things that will probably make the interval tree obsolete. It was fun implementing it, but:

1. Tree sitter set_included_ranges API does not allow overlapping ranges and errors out if you try. It also enforces that the ranges are supplied in order. So we can assume or at least enforce that the ranges are non-overlapping and sorted within each of incoming regions, as there is one tree parse per region and so any overlaps will be caught later anyway. We can also assume, therefore, that this is true of each of the parent's ranges as well.
2. Non-overlapping ranges are easy. You can do a straight O(n) search or, since we are sure they are sorted, a binary search.
3. The main issue is that the parent regions don't line up with the child regions, ie not every parent region will have injections for every child languagetree. This is fixable by making _get_injections return a sparse structure so the indexes line up. Or something to that effect, like storing a parent region index in there.
4. Binary searching for overlapping nodes would give us the logarithmic complexity the interval tree is providing, and N would be smaller. Each search would be on the matching parent region, not the full list of parent ranges. I actually think this will be more correct as well, since unrelated ranges should not be causing injections to clip. It would be a pretty convoluted scenario to have overlapping injections for different languages within the same file, but whatever, let people do what they want.

cormacrelf committed:
fix(tests): don't spuriously initialise api/highlight_spec.lua

cormacrelf committed:
test(treesitter): add some simple included_ranges() tests

cormacrelf committed:
test(treesitter): add regression test for clipping child language regions

cormacrelf committed:
test(treesitter): failing test for cross-region interference

cormacrelf committed:
fix(treesitter): ranges to byte offset for strings

Previously this was using vim.fn.byteidx, which was
not at all doing the job for multiline strings, and
in any case TS ranges use byte offsets already, not
cursor positions.

cormacrelf committed:
fix(treesitter): clip injection ranges to parent language

also passes the region isolation test

This changes the set_included_regions signature but
it is a private API. We need to match each child
region up with the index of the tree that its
injection queries came from, which is (equivalently)
the index of the parent region it came from.
So set_included_regions now takes a sparse array of
Region6[].

cormacrelf committed:
fix(treesitter): add a little null coalescing

cormacrelf commented:
Also a little sneak preview before/after clipping ranges in TSHighlighter as well:

<img width="412" alt="Screen Shot 2022-12-08 at 5 48 51 am" src="https://user-images.githubusercontent.com/378760/206269633-b79812cd-3eef-4111-8795-6dd528e52818.png">

<img width="432" alt="Screen Shot 2022-12-08 at 5 47 45 am" src="https://user-images.githubusercontent.com/378760/206269713-ffbcad3f-c388-4c0f-b611-dde9c7c09e14.png">

cormacrelf commented on runtime/lua/vim/treesitter/languagetree.lua:
```diff
@@ -280,32 +337,22 @@ end
 ---
 ---@private
----@param regions table List of regions this tree should manage and parse.
-function LanguageTree:set_included_regions(regions)
```
A couple of questions:

First, is `set_included_regions` public API or not? It has `@private` luadoc so it isn't in the published docs, but it seems that it used to be. The `@private` was added in #20142. As I was writing this I thought I was free to change the signature and I think I might have to backpedal that.

If it is a public API...

Since we're normalising `Range4`s, that means both the setter and the getter have some undocumented functionality. I'd guess that most people would interpret these APIs as documented as a simple getter and setter combo. But that's already not how it works because we are normalising `Range4`s in-place. So `set_included_regions` currently accepts range4 + tsnode, and `included_regions` returns range6 + tsnode. It looks as though range4 support was added in order to implement the `#offset!` directive, and range6's accidentally leaked out (only on child langs). Out [in the wild][] one or two people seem to be using it a little, entirely with tsnodes and assuming tsnodes so far.

There are a couple of options to rectify this:

* normalise everything to range6 and document that you can supply range4s or range6s when setting and that `included_regions` will return range6 always;
* stop normalising in-place, make it behave like a getter/setter for the root, document that _range4_ works for the setter, pretend range6 in the public API never happened and then, for child langs, re-normalise the return value of `included_regions` back to range4.
* some other permutations of those options

Any ideas? In any case at minimum I will probably normalise tsnodes to range6 internally. We need range6 so we can arbitrarily split them and have a convenient comparison value (the buffer byte offset), and we need uniformity lest the clipping code get really messy,.

[in the wild]: https://github.com/search?type=code&q=set_included_regions+OR+included_regions+language%3Alua+path%3A%2F%5Elua%2F

justinmk replied:
> It has `@private` luadoc so it isn't in the published docs

that means it's private

cormacrelf replied:
I was mainly confused because it didn't have an _underscore prefix like the other private methods. Thanks for clarifying.

justinmk commented:
> * Write a centred interval tree in Lua based on the [Wikipedia outline](https://en.wikipedia.org/wiki/Interval_tree).

This is a lot of code for this specialized purpose, maybe we should expose the datastructure to plugins (can be a followup). Meanwhile, do we need direct test coverage of `_clipping.lua` to make the interface a bit clearer / documented?

cormacrelf commented:
@justinmk I changed the impl to a binary search and deleted the interval tree, should have updated the top comment, but generally yes I think we should make the clipping API public in some capacity. But maybe it should be as an argument to the query iterators, ie you can optionally pass in a region to clip all the matches/captures to. The main reason is that there are some optimisations you can do with it that are probably too finicky for people to figure out themselves.[^1]

And yes it should have its own tests, not only to document the interface but to test some edge cases. Binary searching range overlaps is tricky stuff.

[^1]: They involve setting lower and upper bounds of the search given the ordering, overlapping and boundary characteristics of the three applications (in increasing order of difficulty: clipped LanguageTree child region, clipped query:iter_captures, query:iter_matches).

justinmk commented:
> But maybe it should be as an argument to the query iterators, ie you can optionally pass in a region to clip all the matches/captures to.

That sounds good. Generally, adding a parameter to generic functions is better than adding new special functions.

justinmk commented on runtime/lua/vim/treesitter/languagetree.lua:
```diff
@@ -30,7 +30,7 @@ LanguageTree.__index = LanguageTree
 ---                                runtime file searching for the injection language
 ---                                query per language.
----@return LanguageTree |LanguageTree| parser object
+---@return LanguageTree _ |LanguageTree| parser object
```
?

clason replied:
Needed for lua-language-server to correctly parse this, otherwise it assumes `LanguageTree` is an object of type `|LanguageTree|` (with the pipes), which is breaks parsing since pipes separate multiple optional types.

justinmk commented on runtime/lua/vim/treesitter/languagetree.lua:
```diff
@@ -333,7 +380,10 @@ end
+--
 ---@private
+---@return { [string]: { [number]: Region4[] } }
+---           ^ lang      ^ tree_index (sparse)
```
might want to use `<pre>` here so that it renders better in :help. Hopefully doxygen can deal with this:

```
---@return
--- <pre>
---        { [string]: { [number]: Region4[] } }
---           ^ lang      ^ tree_index (sparse)
--- </pre>

cormacrelf replied:
It's `@private`. It doesn't show up in :help, right?

theHamsta commented:
It's worth investigating whether the PR for extranges can provide infrastructure for this PR. Extranges seem to implement similar functionality as interval trees and could provide functionality like auto-updating/intersection/containment test.

justinmk commented:
> Extranges seem to implement similar functionality as interval trees and could provide functionality like auto-updating/intersection/containment test.

https://github.com/neovim/neovim/pull/13243

cormacrelf commented:
Ok, extranges maybe another time. For now it can clip one range that intersects only a couple of parent ranges in about 600ns when parent_ranges is small. `O(log2(n))` search is pretty flat for a long time (n=#parent_regions), only gets to 750ns/op when n=1000. A decent chunk of that will be simply allocating tiny range6 tables. But it's probably fast enough for now.

The real bottleneck is probably that we are throwing out all child parsers' trees and not reusing them for the next TS parse. One upshot of having a basic range clipper is you can repurpose it to ask "is this tree root contained in this child region?" Combine that with checking if a tree has been made empty (by a `tree:edit()` call), we now have a way to tell which trees can be kept in `_set_included_regions` and which regions to match them up with, striking off that TODO item. So child parsers can now utilise TS incremental parsing.

I've got this going now. Just gotta shape it up to publish.

ayles commented:
Hi! Any progress on this?

justinmk commented:
> > Extranges seem to implement similar functionality as interval trees and could provide functionality like auto-updating/intersection/containment test.
> 
> #13243

extranges PR is now ready for testing.

clason commented:
This has gone stale since our codebase has evolved a lot since this PR was opened, and a more focused solution is possible now (#32549).

Thank you for all the effort you put into this! I like to think that even though it didn't get merged, it was helpful in shaping subsequent development.

cormacrelf commented:
I'm glad to have helped!
